File: DotNetGotchas/CSharp/DisposeFinalize/DisposePattern/Derived.cs
//Derived.cs
using System;

namespace DisposePattern
{
	public class Derived : Base
	{
		private bool disposed = false;

		public Derived(int theID) : base(theID) {}

		protected override void Dispose(bool disposing)
		{
			if (!disposed)
			{
				try
				{
					if (disposing)
					{
						Console.WriteLine(
							"Derived Cleaning up managed resources");
						// Code to clean up managed resources
					}

					Console.WriteLine(
						"Derived Cleaning up unmanaged resources");
					// Code to clean up unmanaged resources
				}
				finally
				{
					base.Dispose(disposing);
				}
			}

			disposed = true;
		}
	}
}

File: DotNetGotchas/VBDN/DisposeFinalize/DisposePattern/Derived.vb
'Derived.vb
Public Class Derived
    Inherits Base

    Private disposed As Boolean = False

    Public Sub New(ByVal theID As Integer)
        MyBase.New(theID)
    End Sub


    Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
        Try
            If Not disposed Then

                If disposing Then
                    Console.WriteLine( _
                        "Derived Cleaning up managed resources")
                    ' Code to clean up managed resources
                End If

                Console.WriteLine( _
                    "Derived Cleaning up unmanaged resources")
                ' Code to clean up unmanaged resources
            End If

            disposed = True
        Finally
            MyBase.Dispose(disposing)
        End Try
    End Sub
End Class
